### PR TITLE
Dynamic ONNX Matmul

### PR DESCRIPTION
**Pull request by CharlieL7**
* Extends `parse_matmul.hpp` to handle dynamic input shapes
* Does not support broadcasting of the outer dimensions for dynamic shapes at this time

### CODE REVIEW

File: src/onnx/parse_matmul.cpp [hunk 1]
```diff
@@ -43,66 +43,127 @@ struct parse_matmul : op_parser<parse_matmul>
...
-            auto output_lens = compute_broadcasted_lens(l0_broadcasted_lens, l1_broadcasted_lens);
-            l0_broadcasted_lens = output_lens;
-            l0_broadcasted_lens.insert(l0_broadcasted_lens.end(), l0_it, l0_lens.end());
-            l1_broadcasted_lens = output_lens;
-            l1_broadcasted_lens.insert(l1_broadcasted_lens.end(), l1_it, l1_lens.end());
-            if(l0_lens != l0_broadcasted_lens)
+            // TODO: handling this case requires a new multibroadcast mode
+            if(not std::equal(
+                   s0_dds.rbegin() + 2, s0_dds.rend(), s1_dds.rbegin() + 2, s1_dds.rend()))
+            {
```
Comment (TedThemistokleous) [inline src/onnx/parse_matmul.cpp:79]: Add a test case for this check.

**Review by TedThemistokleous — CHANGES_REQUESTED** [review]

File: src/onnx/parse_matmul.cpp [hunk 2]
```diff
@@ -43,66 +43,127 @@ struct parse_matmul : op_parser<parse_matmul>
...
+                MIGRAPHX_THROW("PARSE_MATMUL: dynamic shape broadcasting not supported");
+            }
+
+            instruction_ref dot_res = info.add_instruction(make_op(opd.op_name), a0, a1);
+
+            // squeeze the appended or prepended dimensions
+            int64_t num_axis = static_cast<int64_t>(dot_res->get_shape().ndim());
```
Comment (pfultz2) [inline src/onnx/parse_matmul.cpp:86]: The `static_cast` can be removed.

File: src/onnx/parse_matmul.cpp [hunk 3]
```diff
@@ -43,66 +43,127 @@ struct parse_matmul : op_parser<parse_matmul>
...
+                }
+                if(s1_lens != l1_broadcasted_lens)
+                {
+                    ba1 = info.add_instruction(
+                        make_op("multibroadcast", {{"out_lens", l1_broadcasted_lens}}), a1);
+                }
+            }
+            instruction_ref dot_res = info.add_instruction(make_op(opd.op_name), ba0, ba1);
+            // squeeze the appended or prepended dimensions
+            int64_t num_axis = static_cast<int64_t>(dot_res->get_shape().lens().size());
```
Comment (pfultz2) [inline src/onnx/parse_matmul.cpp:152]: The `static_cast` can be removed here as well.

File: src/onnx/parse_matmul.cpp [hunk 4]
```diff
@@ -43,66 +43,127 @@ struct parse_matmul : op_parser<parse_matmul>
...
-        }
-        if(is_b_appended)
+        else
         {
-            dot_res = info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 1}}}), dot_res);
-        }
+            auto s0_lens = s0.lens();
+            auto s1_lens = s1.lens();
+
+            // args[0] is a vector, prepend 1 to the shape
```
Comment (pfultz2) [inline src/onnx/parse_matmul.cpp:106]: We should only prepend 1 when `s0.ndim() != s1.ndim()`.

File: src/onnx/parse_matmul.cpp [hunk 5]
```diff
@@ -43,66 +43,127 @@ struct parse_matmul : op_parser<parse_matmul>
...
+                    info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 2}}}), dot_res);
+                --num_axis;
+            }
+            if(is_b_appended)
+            {
+                dot_res =
+                    info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 1}}}), dot_res);
+            }
+
+            return dot_res;
```
Comment (pfultz2) [inline src/onnx/parse_matmul.cpp:165]: This looks like its the repeating the same thing for dynamic and non-dynamic. This should probably be more unified.

File: src/onnx/parse_matmul.cpp (hunk 4 above)
Comment (CharlieL7) [reply src/onnx/parse_matmul.cpp:106]: I think it's trying to do the same thing as numpy.matmul. The `dot` operator also only works for tensors with `ndim >= 2`.

File: src/onnx/parse_matmul.cpp [hunk 6]
```diff
@@ -43,55 +43,98 @@ struct parse_matmul : op_parser<parse_matmul>
...
+            if(opd.op_name == "quant_dot")
+            {
+                MIGRAPHX_THROW("PARSE_MATMUL: dynamic MatMulInteger not supported");
+            }
+            auto s0_dds = s0.to_dynamic().dyn_dims();
+            auto s1_dds = s1.to_dynamic().dyn_dims();
+            // prepend or append 1 dynamic_dimension for 1D tensors
+            if(is_a_prepended)
+            {
+                s0_dds.insert(s0_dds.begin(), {1, 1});
```
Comment (umangyadav) [inline src/onnx/parse_matmul.cpp:76]: It should just take shape of `a0` and `a1` as `s0` and `s1` which are unsqueezed.

File: src/onnx/parse_matmul.cpp [hunk 7]
```diff
@@ -43,55 +43,98 @@ struct parse_matmul : op_parser<parse_matmul>
...
-            auto l1_it = l1_lens.begin() + l1_lens.size() - 2;
-            std::vector<std::size_t> l1_broadcasted_lens(l1_lens.begin(), l1_it);
-            auto output_lens = compute_broadcasted_lens(l0_broadcasted_lens, l1_broadcasted_lens);
-            l0_broadcasted_lens = output_lens;
-            l0_broadcasted_lens.insert(l0_broadcasted_lens.end(), l0_it, l0_lens.end());
-            l1_broadcasted_lens = output_lens;
-            l1_broadcasted_lens.insert(l1_broadcasted_lens.end(), l1_it, l1_lens.end());
-            if(l0_lens != l0_broadcasted_lens)
+            auto s0_lens = s0.lens();
+            auto s1_lens = s1.lens();
```
Comment (umangyadav) [inline src/onnx/parse_matmul.cpp:95]: Take `s0_lens = a0->shape().lens()`.

**Review by TedThemistokleous — APPROVED** [review]

**Review by umangyadav — APPROVED** [review]

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**